Uncertainty metadata in CF (#661)
cf-convention/cf-conventions: ch10.adoc

The nature of the relationship between a data variable and its __uncertainty variables__ (i.e. those named by its **`uncertainty_variables`** attribute) is determined by the uncertainty variables' attributes.

The provision of uncertainty metadata is designed to be consistent, in both scientific content and terminology, with the "Guide to the expression of uncertainty in measurement" <<GUM>> and the "International Vocabulary of Metrology" <<VIM>>, both published by the Joint Committee for Guides in Metrology, which should be considered the canonical references for this chapter.

Measurement, in the context of uncertainty, is the process of obtaining a numerical value that represents a physical quantity (i.e. the measurand).
A measured value is typically derived from a numerical simulation (e.g. climate model output), a direct observation (e.g. sensor data), or a remote sensing platform (e.g. satellite retrievals).
The error of a measurement is the measured value minus the true, unknowable value of the measurand.
The uncertainty of a measurement is a measure of the dispersion of the values that could reasonably be attributed to the measurand.

> Uncertainty is expressed as a statistical coverage interval (hereafter referred to as a "coverage interval") derived from the probability distribution of the measurand (hereafter referred to as the "probability distribution") that contains the true value with a given probability, i.e. the range of values within which the true value is expected to lie.

Reviewer: Should we make a first numbered heading here, e.g. _10.1 Absolute coverage interval_ ? I find it slightly unsettling that the current 10.1 under "Uncertainty" is "10.1. Relative differences".

> For instance, for a measured value of 50 there could a probability of 0.95 that the true value lies within the coverage interval [40, 60].

Reviewer: a word is missing. ... there could **be** a probability ...

If the uncertainty variable has a **`standard_name`** attribute then it must be identical to the **`standard_name`** attribute of the parent data variable.
However, if the uncertainty variable has no standard name then it may be associated with any appropriate parent data variable with physically equivalent units.
Note that there are different rules regarding the **`units`** and **`standard_name`** attributes of an uncertainty variable that contains relative differences (<<relative-differences>>).

When the coverage interval is symmetrical about the data variable values (i.e. the data variable values lie at the midpoint of the coverage interval), the coverage interval may be fully defined by a single non-negative value: the coverage interval half-width.
This value is equal to the absolute difference between the data variable value and the coverage interval's lower or upper limit.
The coverage interval limits are reconstructed by subtracting the half-width from, and adding it to, the data variable value.

> This case is indicated by specific values of the **`coverage_interval`** attribute (such as `standard_deviation` and `half_width`), for which the size-two interval dimension is not required and must be omitted.

Reviewer: ", for which ..." could refer to "this case", or to "the coverage_interval attribute". Maybe rephrase as:

> This case, for which .... omitted, is indicated .... half_width).

or other solution.

It is deprecated to express uncertainty using an ancillary variable with the standard_error standard name modifier (<<ancillary-data>>).

[[relative-differences, Section 10.1, "Relative differences"]]
=== Relative differences

The coverage interval limits may alternatively be defined as relative differences (i.e. fractional or percentage differences, as opposed to difference offsets) to the data variable values, indicated by an **`uncertainty_representation`** attribute value of `relative_difference`.
For a symmetric coverage interval defined by a single non-negative number (i.e. when the **`coverage_interval`** attribute is `standard_deviation` or `half_width`), the limits are reconstituted by multiplying the data variable value by the factors `1 - f` and `1 + f`, where `f` is the relative offset uncertainty array value expressed as a fraction.
If the uncertainty is provided as a percentage, `f` is found by dividing the uncertainty array value by 100.

> For instance, for a data variable containing a measured value of 20 there could be a probability of 0.95 that the true value lies within the symmetrical coverage interval [15, 25], and the half-width could be stored in an uncertainty variable array as the relative values 25% or 0.25.

Author: Fix typo.
```suggestion
For instance, for a data variable containing a measured value of 20 there could be a probability of 0.95 that the true value lies within the symmetrical coverage interval [15, 25], and the half-width could be stored in an uncertainty variable array as the relative value 25% or 0.25.
```

> For a coverage interval defined by two offsets stored in an interval dimension (i.e. when the **`coverage_interval`** attribute is `offsets`), the limits are reconstituted by multiplying the data variable value by the factors `1 + f` and `1 + g`, where `f` and `g` are the two relative offset uncertainty array values expressed as
> fractions.

Author: Fix one-sentence-per-line text formatting.
```suggestion
For a coverage interval defined by two offsets stored in an interval dimension (i.e. when the **`coverage_interval`** attribute is `offsets`), the limits are reconstituted by multiplying the data variable value by the factors `1 + f` and `1 + g`, where `f` and `g` are the two relative offset uncertainty array values expressed as fractions.
```

`half_width`:: The array has no interval dimension and contains the coverage interval's non-negative half-width.

`offsets`:: The array has a size-two interval dimension that contains the offsets which define the lower and upper limits coverage interval.

==== coverage_probability

The **`coverage_probability`** attribute contains the probability, in the range 0 to 1, that the coverage interval contains the true data variable value.
The attribute must be set. For convenience, <<uncertainty-cv-coverage-probabilities-gaussian, Table N.5>> and  <<uncertainty-cv-coverage-probabilities-uniform, Table N.6>> provide the coverage probabilities for coverage intervals based on the standard uncertainty (i.e. one standard deviation either side of the mean of the distribution), and selected expanded uncertainties, of the Gaussian and uniform distributions.

==== uncertainty_representation

The **`uncertainty_representation`** attribute specifies whether the uncertainty variable array represents absolute or relative differences to the data variable values, and takes one of the following values:

> **`difference`**:: This is the default.

Reviewer: Two comments:

What does _This is the default_ mean here? An uncertainty variable without `:uncertainty_representation` is assumed to have `:uncertainty_representation = "difference"` ? Would it be clearer with _This is the default value._?

Would it be clearer to use `absolute_difference` , for symmetry with the `relative_difference` option ?

[[example-uncertainty-components]]
[caption="Example 10.5. "]
.Uncertainty components
====
----
dimensions:
  time = 30 ;
  lon = 360 ;
  lat = 180 ;

variables:
  // Data variable
  float sst(time, lat, lon) ;
    sst:standard_name = "sea_surface_temperature" ;
    sst:units = "K" ;
    sst:units_metadata = "temperature: on_scale" ;
    sst:uncertainty_variables = "u_total u_random u_systematic" ;
  // Uncertainty variables
  float u_total(time, lat, lon) ;
    u_total:standard_name = "sea_surface_temperature" ;
    u_total:units = "K" ;
    u_total:units_metadata = "temperature: on_scale" ;
    u_total:coverage_interval = "standard_deviation" ;
    u_total:coverage_probability = 0.6827 ;
    u_total:probability_distribution = "gaussian" ;
    u_total:uncertainty_component = "combined_uncertainty" ;
  float u_random(time, lat, lon) ;
    u_random:standard_name = "sea_surface_temperature" ;
    u_random:units = "K" ;
    u_random:units_metadata = "temperature: on_scale" ;
    u_random:coverage_interval = "standard_deviation" ;
    u_random:coverage_factor = 2 ;
    u_random:coverage_probability = 0.9545 ;
    u_random:probability_distribution = "gaussian" ;

>     u_random:uncertainty_component = "Uncorrelated" ;

Reviewer: I find it confusing that `uncertainty_component` is used here for the concept of correlation, since we later introduce a specific framework for describing correlations. Maybe one solution would be to expand the `_component` to make it clear this is free text. For example "Random noise (e.g. from random sensor noise)" here "Systematic uncertainty (e.g. impact of synoptic environment on the retrieval)".

  float u_systematic(lat, lon) ;
    u_systematic:standard_name = "sea_surface_temperature" ;
    u_systematic:units = "K" ;
    u_systematic:units_metadata = "temperature: on_scale" ;
    u_systematic:coverage_interval = "standard_deviation" ;
    u_systematic:coverage_probability = 0.6827 ;
    u_systematic:probability_distribution = "gaussian" ;

>     u_systematic:uncertainty_component = "Correlated over synoptic scales" ;

Reviewer: see my comment [above](https://github.com/cf-convention/cf-conventions/pull/661/changes#r3767427086).

The **`probability_distribution`** attribute is a string attribute of the form: "__distribution [(parameter: value [parameter: value ...])]__", where square brackets indicate optional content.
The "__distribution__" keyword identifies the mathematical form of the probability distribution, and the optional "__parameter: value__" pairs provide parameters that define the distribution.

The "__distribution__" must be one of the probability distribution names defined in <<uncertainty-cv-distribution-type, Table N.1>> (such as `gaussian`).
See <<example-uncertainty-standard-deviation, Example 10.1>>.

A distribution parameter may be defined by a "__parameter: value__" pair.
The type of parameter is named by the "__parameter__" keyword, and must be one of the distribution parameter names defined in <<uncertainty-cv-distribution-parameter, Table N.2>> (such as `skew`).
The "__value__" names a __distribution parameter variable__ whose data array contains the distribution parameter values, and which may have attributes to describe it (such as **`units`** or  **`standard_name`**).
See <<example-uncertainty-distribution, Example 10.6>>.

> A distribution parameter variable must have all or a subset of the dimensions of the uncertainty variable to which it is related.

Author: I think that this should be consistent with error-correlation variables, and be allowed to have any dimension of the parent data variable, even if ones that are not spanned by the uncertainty variable.
```suggestion
A distribution parameter variable must have all or a subset of the dimensions of the parent data variable to which it is related.
```

> The order of dimensions of a distribution parameter variable is not restricted; it does not have to be the same as for the parent uncertainty variable.

Author: uncertainty -> data variable
```suggestion
The order of dimensions of a distribution parameter variable is not restricted; it does not have to be the same as for the parent data variable.
```

> If a moment variable of an uncertainty variable that has been compressed by gathering does not span the compressed dimension (<<compression-by-gathering>>), then the subset of the parent uncertainty variable dimensions is restricted to the uncertainty variable's uncompressed dimensions, i.e. it may have any of the dimensions listed by the compress attribute of the compressed coordinate variable, as well as any of the other dimensions of the uncertainty variable (those apart from the compressed dimension).

Author: uncertainty -> data variable
```suggestion
If a moment variable of an uncertainty variable that has been compressed by gathering does not span the compressed dimension (<<compression-by-gathering>>), then the subset of the parent data variable dimensions is restricted to the data variable's uncompressed dimensions, i.e. it may have any of the dimensions listed by the compress attribute of the compressed coordinate variable, as well as any of the other dimensions of the data variable (those apart from the compressed dimension).
```

[[example-uncertainty-distribution]]
[caption="Example 10.6. "]
.Uncertainty probability distribution
====
----
dimensions:
  time = 30 ;
  interval_offset = 2 ;

variables:
  // Data variable
  float q(time) ;
    q:standard_name = "specific_humidity" ;
    q:units = "g/g" ;
    q:uncertainty_variables = "q_uncertainty" ;
  // Uncertainty variable
  float q_uncertainty(time, interval_offset) ;
    q_uncertainty:standard_name = "specific_humidity" ;
    q_uncertainty:units = "g/g" ;
    q_uncertainty:coverage_interval = "offsets" ;
    q_uncertainty:coverage_probability = 0.9 ;
    q_uncertainty:probability_distribution = "skewed_gaussian (skew: skew kurtosis: kurtosis)" ;
  // Distribution parameter variables
  float skew(time) ;
    skew:units = "1" ;
  float kurtosis(time) ;
    kurtosis:units = "1" ;
----
The uncertainty variable `q_uncertainty` contains a coverage interval defined by lower and upper limits.
The probability that the true values of the data variable `q` are within the coverage interval limits is 0.9.

> The errors are characterized by a skew-normal distribution, which is further defined at each point with its skewness (i.e. the third standardized statistical moment), that is provided in a separate distribution parameter variable.

Reviewer: for clarity, also mention kurtosis in the caption of the Example, since it is also specified?

Since the data variable is `q`, and the uncertainty variable is `q_uncertainty`, maybe it would be more intuitive to use `q_skew` and `q_kurtosis` as variable names? (of course, CF does not work from variable names).

====

[[error-correlation, Section 10.4, "Error correlation"]]
=== Error correlation

The uncertainty variable array provides a measure of the magnitude of the error in the data variable at each domain location.
The error-correlations describe the pattern, independently of the magnitude of the errors themselves, of statistical relationships between the unknown actual errors at different domain locations.
For instance, this information can be used for assessing the uncertainty of derived products (such as those formed by averaging, or by subtraction of a statistical norm), and for the weighting of observations when updating numerical models via data assimilation systems.

Error correlation information may be defined by the uncertainty variable's **`error_correlation`** attribute, which relates data variable dimensions to matrices of error-correlation coefficients (possibly in parametric form), or to a non-standardized description of the error-correlation structure, or both.

> In particular, the parametric form of an error-correlation matrix always names the structural type of the correlations (such as uncorrelated or fully_correlated), providing physical insight into the error structure.

Reviewer: he underscore in fully_correlated gives the impression of a controlled vocabulary, however the controlled vocabulary is `common`. Either use the control vocabulary here, or remove the underscore.

The allowed parametric forms are defined in <<uncertainty-cv>>.

Error-correlation information may be given for any dimensions of the data variable, including any that are not spanned by the uncertainty variable.
This is because, even though the probability distribution characterized by the uncertainty variable does not depend on such dimensions, the error correlations along them may still have different characteristics to those of other dimensions.

> The **`error_correlation`** attribute may be used to describe the correlations of errors across the domain.

Reviewer: This paragraph reads like a re-start of the section. Maybe delete the first sentence, and move the next two into the first paragraph of the section? Especially that the attribute is optional and there is no default value should come earlier.

[[example-uncertainty-error-correlation-1a-2a]]
[caption="Example 10.7. "]
.One-dimensional error correlations
====
----
dimensions:
  time = 30 ;
  lon = 360 ;
  lat = 180 ;
  lat1 = 180 ;

variables:
  // Data variable
  float sst(time, lat, lon) ;
    sst:standard_name = "sea_surface_temperature" ;
    sst:units = "K" ;
    sst:units_metadata = "temperature: on_scale" ;
    sst:uncertainty_variables = "unc" ;
  // Uncertainty variable
  float unc(lat, lon) ;
    unc:standard_name = "sea_surface_temperature" ;
    unc:units = "K" ;
    unc:units_metadata = "temperature: on_scale" ;
    unc:coverage_interval = "standard_deviation" ;
    unc:coverage_probability = 0.6827 ;
    unc:probability_distribution = "gaussian" ;
    unc:error_correlation = "lat: correlations
                             time: uncorrelated
                             lon: triangular (localization_radius: 15)" ;
  // Error-correlation variable
  float correlations(lat, lat1) ;
----

Error correlations described by error-correlation matrices that demonstrate forms 1a, 2a, and 2b of the **`error_correlation`** attribute.
Other uncertainty attributes have been omitted for clarity.

> For the `lat` dimension there is an error-correlation variable called correlations giving actual the error-correlation coefficients (form 1a).

Reviewer: fix font: named `correlations`

The error-correlation matrix is square and indexes the given dimensions after they have been flattened in row-major order, i.e. the first of the dimensions given in the **`error_correlation`** attribute varies the slowest, and the last varies the fastest.
The number of rows or columns in the matrix is therefore equal to the product of the sizes of the given dimensions.
For instance, if the dimension `lat` has size 20 and dimension `lon` has size 30, then `lat: lon: varname` means that the error-correlation variable `varname` contains a 600 by 600 error-correlation matrix for the `lat` and `lon` dimensions that is composed of 400 non-overlapping square blocks, each of size 30 by 30.
Each block along the main diagonal represents the error-correlations between pairs of cells on the `lon` dimension at a specific `lat` dimension location, and the off-diagonal blocks represent the correlations between pairs of cells on the `lon` dimension that have different `lat` dimension locations.
See <<example-uncertainty-error-correlation-1b, Example 10.8>>.

> To save space in the dataset, form 1b should only be used when the error-correlation structure is intrinsically coupled across the dimensions, and cannot be calculated using the Kronecker product <<KRONECKER>> of error-correlation matrices that each apply to a single dimension, which are defined by form 1a.

Author: Singling out 1-d error-correlation matrices is too restrictive. If a 3-d error correlation matrix can be formed by the Kronecker product of matrices for 1 dimension and for 2 dimensions, then that is good.
```suggestion
To save space in the dataset, form 1b should only be used when the error-correlation structure is intrinsically coupled across the dimensions, and cannot be calculated using the Kronecker product <<KRONECKER>> of error-correlation matrices that apply to non-overlapping subsets of the named dimensions.
```

> To save space in the dataset, form 1b should only be used when the error-correlation structure is intrinsically coupled across the dimensions, and cannot be calculated using the Kronecker product <<KRONECKER>> of error-correlation matrices that each apply to a single dimension, which are defined by form 1a.

Reviewer: How would one use the Kronecker product in practice? Is the Kronecker product of correlation matrices along single dimensions the same as form 1a? What about if we had to encode the Kronecker-combination of a matrix defined along 2 dimensions (form 1b) with another along one dimension (form 1a).

I am not 100% sure I understand the implication of the Kronecker product here, so I do not feel I can propose a solution.

[caption="Example 10.8. "]
.Multi-dimensional error correlations
====
----
dimensions:
  time = 30 ;
  lon = 360 ;
  lat = 180 ;
  lat_lon = 64800 ;
  lat_lon1 = 64800 ;

variables:
  // Data variable
  float sst(time, lat, lon) ;
    sst:standard_name = "sea_surface_temperature" ;
    sst:units = "K" ;
    sst:units_metadata = "temperature: on_scale" ;
    sst:uncertainty_variables = "unc" ;
  // Uncertainty variable
  float unc(time, lat, lon) ;
    unc:standard_name = "sea_surface_temperature" ;
    unc:units = "K" ;
    unc:units_metadata = "temperature: on_scale" ;
    unc:coverage_interval = "standard_deviation" ;
    unc:coverage_probability = 0.6827 ;
    unc:error_correlation = "lat: lon: correlations" ;
  // Error-correlation variable
  float correlations(lat_lon, lat_lon1) ;
----

Error correlations described by an error-correlation matrix that demonstrates form 1b of the **`error_correlation`** attribute.
For the `lat` and `lon` dimensions combined there is a single error-correlation variable called `correlations`. This variable's data is a 64800 by 64800 matrix, where `64800 = 180 x 360`, i.e. the product of the sizes of the `lat` and `lon` dimensions.

> Note that error-correlation information does not need to be provided for every data variable dimension, as is this case here.

Reviewer: typo? "as is THE case here"

====

*Form 2a* is a parameterized alternative to forms 1a and 1b, for which "__error_correlation_structure__" names an error-correlation structural type which implies the error-correlation coefficient matrix for every pair of cells indexed by the given dimension or dimensions.

> The structural type must be one of the those defined in <<uncertainty-cv-structural-type, Table N.3>>.

Reviewer: typo, "must be one of ~~the~~ those"
